File: Motocitizen/src/motocitizen/fragments/DetailVolunteersFragment.java
package motocitizen.fragments;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.DialogFragment;
import android.app.Fragment;
import android.app.FragmentTransaction;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;

import motocitizen.activity.AccidentDetailsActivity;
import motocitizen.content.Content;
import motocitizen.content.accident.Accident;
import motocitizen.content.volunteer.VolunteerAction;
import motocitizen.main.R;
import motocitizen.network.requests.AccidentListRequest;
import motocitizen.network.requests.CancelOnWayRequest;
import motocitizen.network.requests.OnWayRequest;
import motocitizen.rows.RowFactory;
import motocitizen.user.User;
import motocitizen.utils.Preferences;

public class DetailVolunteersFragment extends AccidentDetailsFragments {

    private static final int DIALOG_ONWAY_CONFIRM        = 1;
    private static final int DIALOG_ACC_NOT_ACTUAL       = 2;
    private static final int DIALOG_CANCEL_ONWAY_CONFIRM = 3;

    private ImageButton onwayButton;
    private ImageButton onwayCancelButton;
    private ImageButton onwayDisabledButton;
    private View        onwayContent;

    // TODO: Rename and change types and number of parameters
    public static DetailVolunteersFragment newInstance(int accID) {
        DetailVolunteersFragment fragment = new DetailVolunteersFragment();
        Bundle                   args     = new Bundle();
        args.putInt(ACCIDENT_ID, accID);
        fragment.setArguments(args);
        return fragment;
    }

    public DetailVolunteersFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View viewMain = inflater.inflate(R.layout.fragment_detail_volunteers, container, false);

        View toMap = viewMain.findViewById(R.id.details_to_map_button);
        onwayButton = (ImageButton) viewMain.findViewById(R.id.onway_button);
        onwayCancelButton = (ImageButton) viewMain.findViewById(R.id.onway_cancel_button);
        onwayDisabledButton = (ImageButton) viewMain.findViewById(R.id.onway_disabled_button);
        onwayContent = viewMain.findViewById(R.id.acc_onway_table);

        onwayDisabledButton.setEnabled(false);

        onwayButton.setOnClickListener(v -> showDialog(DIALOG_ONWAY_CONFIRM));
        onwayCancelButton.setOnClickListener(v -> showDialog(DIALOG_CANCEL_ONWAY_CONFIRM));
        toMap.setOnClickListener(v -> ((AccidentDetailsActivity) getActivity()).jumpToMap());

//        update();

        return viewMain;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        update();
    }

    private void update() {
        Accident accident = ((AccidentDetailsActivity) getActivity()).getCurrentPoint();

        if (accident == null) {
            showDialog(DIALOG_ACC_NOT_ACTUAL);
            return;
        }

        setupAccess();
        ViewGroup vg_onway = (ViewGroup) onwayContent;
        vg_onway.removeAllViews();
        for (VolunteerAction v : accident.getVolunteers()) {
            vg_onway.addView(RowFactory.INSTANCE.make(getActivity(), v));
        }
    }

    private void setupAccess() {
        Accident accident = ((AccidentDetailsActivity) getActivity()).getCurrentPoint();
        int      id       = accident.getId();
        boolean  active   = accident.isActive() && User.getInstance(getActivity()).isAuthorized();
        onwayButton.setVisibility(id != Preferences.Companion.getInstance(getActivity()).getOnWay() && id != Content.INSTANCE.getInPlace() && active ? View.VISIBLE : View.GONE);
        onwayCancelButton.setVisibility(id == Preferences.Companion.getInstance(getActivity()).getOnWay() && id != Content.INSTANCE.getInPlace() && active ? View.VISIBLE : View.GONE);
        onwayDisabledButton.setVisibility(id == Content.INSTANCE.getInPlace() && active ? View.VISIBLE : View.GONE);
    }

    @SuppressLint("CommitTransaction")
    private void showDialog(int type) {
        mStackLevel++;
        FragmentTransaction ft   = getActivity().getFragmentManager().beginTransaction();
        Fragment            prev = getActivity().getFragmentManager().findFragmentByTag("dialog");
        if (prev != null) {
            ft.remove(prev);
        }
        ft.addToBackStack(null);
        ft.commit();
        Activity act = getActivity();

        switch (type) {
            case DIALOG_ONWAY_CONFIRM:
                DialogFragment onwayConfirm = ConfirmDialog.newInstance(act.getString(R.string.title_dialog_onway_confirm), act.getString(android.R.string.yes), act.getString(android.R.string.no));
                onwayConfirm.setTargetFragment(this, type);
                onwayConfirm.show(getFragmentManager().beginTransaction(), "dialog");
                break;

            case DIALOG_CANCEL_ONWAY_CONFIRM:
                DialogFragment cancelOnwayConfirm = ConfirmDialog.newInstance(act.getString(R.string.title_dialog_cancel_onway_confirm),
                                                                              act.getString(android.R.string.yes),
                                                                              act.getString(android.R.string.no));
                cancelOnwayConfirm.setTargetFragment(this, type);
                cancelOnwayConfirm.show(getFragmentManager().beginTransaction(), "dialog");
                break;

            case DIALOG_ACC_NOT_ACTUAL:
                DialogFragment dialogFrag = ConfirmDialog.newInstance(act.getString(R.string.title_dialog_acc_not_actual), act.getString(android.R.string.ok), "");
                dialogFrag.setTargetFragment(this, type);
                dialogFrag.show(getFragmentManager().beginTransaction(), "dialog");
                break;
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case DIALOG_ONWAY_CONFIRM:
                if (resultCode == Activity.RESULT_OK) {
                    sendOnWay();
                } else if (resultCode == Activity.RESULT_CANCELED) {
                    // todo After Cancel code.
                }
                break;
            case DIALOG_CANCEL_ONWAY_CONFIRM:
                if (resultCode == Activity.RESULT_OK) {
                    sendCancelOnWay();
                } else if (resultCode == Activity.RESULT_CANCELED) {
                    // todo After Cancel code.
                }
                break;
            case DIALOG_ACC_NOT_ACTUAL:
                getActivity().finish();
                break;
        }
    }

    private void sendOnWay() {
        Preferences.Companion.getInstance(getActivity()).setOnWay(accidentID);
        new OnWayRequest(accidentID, response -> new AccidentListRequest(result -> {
            Content.INSTANCE.requestUpdate();
            getActivity().runOnUiThread(() -> {
                ((AccidentDetailsActivity) getActivity()).update();
                update();
            });
        }));
    }

    private void sendCancelOnWay() {
        Preferences.Companion.getInstance(getActivity()).setOnWay(0);
        new CancelOnWayRequest(accidentID, response -> new AccidentListRequest(result -> {
            Content.INSTANCE.requestUpdate();
            getActivity().runOnUiThread(() -> {
                ((AccidentDetailsActivity) getActivity()).update();
                update();
            });
        }));
    }
}

File: Motocitizen/src/motocitizen/fragments/DetailVolunteersFragment.kt
package motocitizen.fragments

import android.app.Activity
import android.app.Fragment
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import motocitizen.activity.AccidentDetailsActivity
import motocitizen.content.Content
import motocitizen.content.accident.Accident
import motocitizen.main.R
import motocitizen.network.CoreRequest
import motocitizen.network.requests.CancelOnWayRequest
import motocitizen.network.requests.OnWayRequest
import motocitizen.rows.volunteer.VolunteerRowFactory
import motocitizen.user.User
import motocitizen.utils.Preferences
import org.json.JSONObject

class DetailVolunteersFragment() : Fragment() {
    private val ROOT_LAYOUT = R.layout.fragment_detail_volunteers
    private val CONTENT_VIEW = R.id.acc_onway_table
    private val TO_MAP_BUTTON = R.id.details_to_map_button
    private val CONFIRM_BUTTON = R.id.onway_button
    private val CANCEL_BUTTON = R.id.onway_cancel_button
    private val DISABLED_BUTTON = R.id.onway_disabled_button

    private val DIALOG_ON_WAY_CONFIRM = 1
    private val DIALOG_CANCEL_ON_WAY_CONFIRM = 2

    private lateinit var confirmButton: ImageButton
    private lateinit var cancelButton: ImageButton
    private lateinit var disabledButton: ImageButton
    private lateinit var content: ViewGroup

    private lateinit var accident: Accident

    constructor(accident: Accident) : this() {
        this.accident = accident
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(ROOT_LAYOUT, container, false)

        val toMapButton = rootView.findViewById(TO_MAP_BUTTON)
        confirmButton = rootView.findViewById(CONFIRM_BUTTON) as ImageButton
        cancelButton = rootView.findViewById(CANCEL_BUTTON) as ImageButton
        disabledButton = rootView.findViewById(DISABLED_BUTTON) as ImageButton
        content = rootView.findViewById(CONTENT_VIEW) as ViewGroup

        disabledButton.isEnabled = false

        confirmButton.setOnClickListener { _ -> showOnWayDialog() }
        cancelButton.setOnClickListener { _ -> showCancelDialog() }
        toMapButton.setOnClickListener { _ -> (activity as AccidentDetailsActivity).jumpToMap() }

        return rootView
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        update()
    }

    private fun update() {
        setupAccess()
        content.removeAllViews()
        accident.volunteers.forEach { action -> content.addView(VolunteerRowFactory.make(activity, action)) }
    }

    private fun setupAccess() {
        with(accident) {
            val active = accident.isActive() && User.isAuthorized
            confirmButton.visibility = if (id != Preferences.onWay && id != Content.inPlace && active) View.VISIBLE else View.GONE
            cancelButton.visibility = if (id == Preferences.onWay && id != Content.inPlace && active) View.VISIBLE else View.GONE
            disabledButton.visibility = if (id == Content.inPlace && active) View.VISIBLE else View.GONE
        }
    }

    private fun showOnWayDialog() {
        val onWayConfirm = ConfirmDialog(activity.getString(R.string.title_dialog_onway_confirm))
        onWayConfirm.setTargetFragment(this, DIALOG_ON_WAY_CONFIRM)
        onWayConfirm.show(fragmentManager, "dialog")
    }

    private fun showCancelDialog() {
        val cancelOnWayConfirm = ConfirmDialog(activity.getString(R.string.title_dialog_cancel_onway_confirm))
        cancelOnWayConfirm.setTargetFragment(this, DIALOG_CANCEL_ON_WAY_CONFIRM)
        cancelOnWayConfirm.show(fragmentManager, "dialog")
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        if (resultCode == Activity.RESULT_CANCELED) return
        when (requestCode) {
            DIALOG_ON_WAY_CONFIRM        -> sendOnWay()
            DIALOG_CANCEL_ON_WAY_CONFIRM -> sendCancelOnWay()
        }
    }

    private fun sendOnWay() {
        Preferences.onWay = accident.id
        OnWayRequest(accident.id, object : CoreRequest.RequestResultCallback {
            override fun call(response: JSONObject) {
                //todo
            }
        })
    }

    private fun sendCancelOnWay() {
        Preferences.onWay = 0
        CancelOnWayRequest(accident.id, object : CoreRequest.RequestResultCallback {
            override fun call(response: JSONObject) {
                //todo
            }
        })
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(AccidentDetailsActivity.ACCIDENT_ID_KEY, accident.id)
        super.onSaveInstanceState(outState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState == null) return
        accident = Content.accidents[savedInstanceState.getInt(AccidentDetailsActivity.ACCIDENT_ID_KEY)]!!
    }
}
